lms/djangoapps/class_dashboard/dashboard_data.py

```diff
@@ -102,59 +162,87 @@ def get_problem_set_grade_distrib(course_id, problem_set):
 
     `problem_set` an array of UsageKeys representing problem module_id's.
 
+    `enrollment` the number of students enrolled in this course.
+
     Requests from the database the a count of each grade for each problem in the `problem_set`.
 
     Returns a dict, where the key is the problem 'module_id' and the value is a dict with two parts:
       'max_grade' - the maximum grade possible for the course
       'grade_distrib' - array of tuples (`grade`,`count`) ordered by `grade`
     """
-
-    # Aggregate query on studentmodule table for grade data for set of problems in course
-    db_query = models.StudentModule.objects.filter(
-        course_id__exact=course_id,
-        grade__isnull=False,
-        module_type__in=PROB_TYPE_LIST,
-        module_state_key__in=problem_set,
-    ).values(
-        'module_state_key',
-        'grade',
-        'max_grade',
-    ).annotate(count_grade=Count('grade')).order_by('module_state_key', 'grade')
-
     prob_grade_distrib = {}
 
-    # Loop through resultset building data for each problem
-    for row in db_query:
-        row_loc = course_id.make_usage_key_from_deprecated_string(row['module_state_key'])
-        if row_loc not in prob_grade_distrib:
-            prob_grade_distrib[row_loc] = {
-                'max_grade': 0,
-                'grade_distrib': [],
-            }
+    if enrollment <= settings.MAX_ENROLLEES_FOR_METRICS_USING_DB:
+        # Aggregate query on studentmodule table for grade data for set of problems in course
+        queryset = models.StudentModule.objects.filter(
+            course_id__exact=course_id,
+            grade__isnull=False,
+            module_type__in=PROB_TYPE_LIST,
+            module_state_key__in=problem_set,
+        ).values(
+            'module_state_key',
+            'grade',
+            'max_grade',
+        ).annotate(count_grade=Count('grade')).order_by('module_state_key', 'grade')
+
+        # Loop through resultset building data for each problem
+        for row in queryset:
+            row_loc = course_id.make_usage_key_from_deprecated_string(row['module_state_key'])
```

Comment:
From the mentality of "code will be read many more times than it will be written", I tend to prefer more verbose variable names.

What is `loc`; is it a `location`, a `locator`, a `locale`, or is it something else?

Sure, we can use context clues in this example, but it's not always evident and explicit naming conventions remove this ambiguity.

Even something like `usage_key = course_id.make_usage_key_from_deprecated_string(...)` makes this more clear.

autocomplete ftw :)